### PR TITLE
[12.0][ADD] graphql_base and graphql_demo

### PR DESCRIPTION
**Pull request by sbidoul**
TODO
- [x] test error handling
- [x] add test in graphql_demo
- [x] some documentation
- [x] better way to monkey patch for json input?

To test this, go to runbot, and replace the URL path with `/graphiql/demo`, and start playing.

peek 2018-11-19 14-51

This animation is based on then demo schema.py in the `graphql_demo`.

image

### CODE REVIEW

File: graphql_base/controllers/main.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            encode=partial(json_encode, pretty=False),
+        )
+        # TODO what to do with status_code?
+        return http.request.make_response(
+            result, headers={"Content-Type": "application/json"}
+        )
+
+    def _handle_graphql_request(self, schema):
+        data = self._parse_body()
+        return self._process_request(schema, data, catch=False)
```
Comment (sbidoul) [inline graphql_base/controllers/main.py]: TODO: to catch or not to catch?

**Comment by sbidoul** [comment]
Runbot is behaving weirdly. When I try to open this PR, I'm redirected to OCB. Has anyone ever seen that behaviour?

**Comment by lmignon** [comment]
@sbidoul The demo addon seems to work properly on runbot 

image

**Comment by sbidoul** [comment]
@lmignon thanks, runbot looks ok now.

**Comment by sbidoul** [comment]
I deem this ready for review.

File: graphql_demo/tests/test_graphene.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        # disable logging for the graphql executor because we are testing
+        # errors and OCA's test runner considers the two errors being logged
+        # as fatal
+        logging.getLogger("graphql.execution").setLevel(logging.CRITICAL)
+
+    def execute(self, query):
+        res = self.client.execute(query, context={"env": self.env})
+        if not res:
+            raise RuntimeError("GarphQL query returned no data")
+        if res.get("error"):
```
Comment (sbejaoui) [inline graphql_demo/tests/test_graphene.py]: I think the key is errors?

Comment (sbidoul) [reply graphql_demo/tests/test_graphene.py]: @sbejaoui good catch. This is fixed now.

**Review by lmignon — APPROVED** [review]
Great! (Code review + functional tests)

**Comment by sbidoul** [comment]
@sbejaoui can you update your review?

**Review by sbejaoui — APPROVED** [review]

**Comment by lmignon** [comment]
@sbidoul If OK for you I'll merge.

**Comment by sbidoul** [comment]
Yes go ahead with merging. We are using it successfully on a large project.

**Pull request merged**